### PR TITLE
refactor: Do not reconcile on any error inside finalizers

### PR DESCRIPTION
**Pull request by mmorhun**
Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

### What does this PR do?
Refactors component's finalizers in order to not to hung Che deletion if an error happen in one of the finalizers. 
New behavior logs all errors in all fnalizers, but do not stop current reconcile loop. If at least one error happened at least in one finalizer, a new reconcile loop is queued.

### Screenshot/screencast of this PR
N/A

### What issues does this PR fix or reference?
N/A

### How to test this PR?
N/A, refactoring

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [ ] [OLM bundles are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: pkg/deploy/reconcile_manager.go [hunk 1]
```diff
@@ -66,10 +66,10 @@ func (manager *ReconcileManager) ReconcileAll(ctx *DeployContext) (reconcile.Res
 
 func (manager *ReconcileManager) FinalizeAll(ctx *DeployContext) {
```
Comment (tolusha) [inline pkg/deploy/reconcile_manager.go:67]: We must return (done bool) to trigger a reconcile again if any Finalizer failed

File: pkg/deploy/dashboard/dashboard.go [hunk 2]
```diff
@@ -103,18 +104,20 @@ func (d *DashboardReconciler) Reconcile(ctx *deploy.DeployContext) (reconcile.Re
 	return reconcile.Result{}, true, nil
 }
 
-func (d *DashboardReconciler) Finalize(ctx *deploy.DeployContext) error {
-	done, err := deploy.Delete(ctx, types.NamespacedName{Name: d.getClusterRoleName(ctx)}, &rbacv1.ClusterRole{})
-	if !done {
-		return err
+func (d *DashboardReconciler) Finalize(ctx *deploy.DeployContext) bool {
+	if _, err := deploy.Delete(ctx, types.NamespacedName{Name: d.getClusterRoleName(ctx)}, &rbacv1.ClusterRole{}); err != nil {
```
Comment (tolusha) [inline pkg/deploy/dashboard/dashboard.go:108]:
```suggestion
       done := true
	if _, err := deploy.Delete(ctx, types.NamespacedName{Name: d.getClusterRoleName(ctx)}, &rbacv1.ClusterRole{}); err != nil {
```

File: pkg/deploy/dashboard/dashboard.go [hunk 3]
```diff
@@ -103,18 +104,20 @@ func (d *DashboardReconciler) Reconcile(ctx *deploy.DeployContext) (reconcile.Re
...
+		logrus.Errorf("Failed to delete ClusterRole %s, cause: %v", d.getClusterRoleName(ctx), err)
```
Comment (tolusha) [inline pkg/deploy/dashboard/dashboard.go:109]:
```suggestion
                done = false
		logrus.Errorf("Failed to delete ClusterRole %s, cause: %v", d.getClusterRoleName(ctx), err)
```

File: pkg/deploy/dashboard/dashboard.go [hunk 4]
```diff
@@ -103,18 +104,20 @@ func (d *DashboardReconciler) Reconcile(ctx *deploy.DeployContext) (reconcile.Re
...
 	}
 
-	done, err = deploy.Delete(ctx, types.NamespacedName{Name: d.getClusterRoleBindingName(ctx)}, &rbacv1.ClusterRoleBinding{})
-	if !done {
-		return err
+	if _, err := deploy.Delete(ctx, types.NamespacedName{Name: d.getClusterRoleBindingName(ctx)}, &rbacv1.ClusterRoleBinding{}); err != nil {
+		logrus.Errorf("Failed to delete ClusterRoleBinding %s, cause: %v", d.getClusterRoleBindingName(ctx), err)
```
Comment (tolusha) [inline pkg/deploy/dashboard/dashboard.go:113]:
```suggestion
                done = false
		logrus.Errorf("Failed to delete ClusterRoleBinding %s, cause: %v", d.getClusterRoleBindingName(ctx), err)
```

File: pkg/deploy/dashboard/dashboard.go [hunk 5]
```diff
@@ -103,18 +104,20 @@ func (d *DashboardReconciler) Reconcile(ctx *deploy.DeployContext) (reconcile.Re
...
 	}
 
-	return deploy.DeleteFinalizer(ctx, ClusterPermissionsDashboardFinalizer)
+	if err := deploy.DeleteFinalizer(ctx, ClusterPermissionsDashboardFinalizer); err != nil {
+		logrus.Errorf("Error deleting finalizer: %v", err)
+		return false
```
Comment (tolusha) [inline pkg/deploy/dashboard/dashboard.go:118]:
```suggestion
                done = false
		logrus.Errorf("Error deleting finalizer: %v", err)
```

File: pkg/deploy/dashboard/dashboard.go [hunk 6]
```diff
@@ -103,18 +104,20 @@ func (d *DashboardReconciler) Reconcile(ctx *deploy.DeployContext) (reconcile.Re
...
+	}
+	return true
```
Comment (tolusha) [inline pkg/deploy/dashboard/dashboard.go:120]:
```suggestion
	return done
```

File: controllers/che/checluster_controller.go [hunk 7]
```diff
@@ -319,10 +319,9 @@ func (r *CheClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 			return ctrl.Result{}, nil
 		}
 	} else {
-		r.reconcileManager.FinalizeAll(deployContext)
+		done := r.reconcileManager.FinalizeAll(deployContext)
+		return ctrl.Result{Requeue: done}, nil
```
Comment (tolusha) [inline controllers/che/checluster_controller.go:323]:
```suggestion
		return ctrl.Result{Requeue: !done}, nil
```

**Review by tolusha — APPROVED** [review]

File: controllers/che/checluster_controller.go (hunk 7 above)
Comment (mmorhun) [reply controllers/che/checluster_controller.go]: Good catch, thank you!

**Review by AndrienkoAleksandr — APPROVED** [review]

**Pull request merged**